File: cms-app-crawler-js/src/base/Locator.js
'use strict';

class Locator {
  constructor(type, value) {
    this.type = type;
    this.value = value;
    Object.freeze(this);
  }

  getSelectorKey() {
    switch (this.type) {
      case 'id':          return '#' + this.value;
      case 'name':        return `[name='${this.value}']`;
      case 'xpath':       return 'xpath=' + this.value;
      case 'data-testid': return 'data-testid=' + this.value;
      case 'aria-label':  return 'aria-label=' + this.value;
      default:            return this.value;
    }
  }

  toPlaywrightSelector() {
    switch (this.type) {
      case 'id':          return `#${this.value}`;
      case 'name':        return `[name='${this.value}']`;
      case 'xpath':       return `xpath=${this.value}`;
      case 'data-testid': return `[data-testid='${this.value}']`;
      case 'aria-label':  return `[aria-label='${this.value}']`;
      default:            return this.value;
    }
  }
}

module.exports = { Locator };

File: cms-app-crawler-py/src/base/locator.py
from dataclasses import dataclass


@dataclass(frozen=True)
class Locator:
    type: str  # "css" | "xpath" | "id" | "name" | "data-testid" | "aria-label"
    value: str

    @property
    def selector_key(self) -> str:
        match self.type:
            case "id":          return f"#{self.value}"
            case "name":        return f"[name='{self.value}']"
            case "xpath":       return f"xpath={self.value}"
            case "data-testid": return f"data-testid={self.value}"
            case "aria-label":  return f"aria-label={self.value}"
            case _:             return self.value

    @property
    def playwright_selector(self) -> str:
        match self.type:
            case "id":          return f"#{self.value}"
            case "name":        return f"[name='{self.value}']"
            case "xpath":       return f"xpath={self.value}"
            case "data-testid": return f"[data-testid='{self.value}']"
            case "aria-label":  return f"[aria-label='{self.value}']"
            case _:             return self.value
